# Portfolio assignment 21
**Train a perceptron to predict one of the categorical columns of your own dataset(Pokemon.csv):**

- Prepare the data
- Some machine learning algorithms can not handle categorical values. You will either need to: To handle categorical data, you can use One-Hot Encoding to convert categories into binary columns with .get_dummies(<value(s)>). This creates a new column for each category
- Split your dataset into training (70%) and testing (30%) sets.
- Use your Perceptron to make predictions for both the train and test set.

---

- Fit a Perceptron model using your own selected feature columns.
- Fine-tune the learning rate and epochs (max iterations).
- Calculate the accuracy for both the train set predictions and test set predictions, what happens per iteration?
- Is the accurracy different? Did you expect this difference?

---


In [7]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Perceptron
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import OneHotEncoder

pokemon = pd.read_csv("pokemon.csv")
pokemon = pokemon.dropna()

# Doelvariabele (Type 1 voorspellen)
# Groepeer types: 'Fire', 'Water', 'Grass' → 'Elemental', rest → 'Other' ( 11% > 70% )
y = pokemon['Type_Group'] = pokemon['Type 1'].apply(lambda x: 'Elemental' if x in ['Fire', 'Water', 'Grass'] else 'Other')
# Numerieke features
X = pokemon[["Attack", "Defense", "Speed", "HP"]]


from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
y_encoded = le.fit_transform(y)

X_train, X_test, y_train, y_test = train_test_split(X, y_encoded, test_size=0.3, random_state=42)

In [9]:
model = Perceptron(max_iter=1000, eta0=1.0, random_state=42)
model.fit(X_train, y_train)

# Voorspellingen
y_pred_train = model.predict(X_train)
y_pred_test = model.predict(X_test)

# Accuracy
train_acc = accuracy_score(y_train, y_pred_train)
test_acc = accuracy_score(y_test, y_pred_test)

print("Train accuracy:", round(train_acc, 2))
print("Test accuracy:", round(test_acc, 2))

Train accuracy: 0.71
Test accuracy: 0.75


---

### Findings:
Doel: origineel "type 1" maar dit heb ik gegroepeerd naar 3 types in "Elemental" (Fire, Water, Grass) vs "Other" omdat de accuracy dan hoger uitkomt (voorheen maar 12%)
Model: Perceptron met numerieke features Attack, Defense, Speed, HP

#### Prestaties:
- Train accuracy: 71%
- Test accuracy: 75%

**Findings:**
- Het model presteert goed met simpele kenmerken
- De accuracies liggen dicht bij elkaar → geen overfitting
- Simpele lineaire scheiding (Perceptron) is hier voldoende